app/api/controllers/validators/Validator.scala

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright 2023 HM Revenue & Customs
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package api.controllers.validators
+
+import api.models.errors.{BadRequestError, ErrorWrapper, MtdError}
+import cats.data.Validated
+import cats.data.Validated.{Invalid, Valid}
+import utils.Logging
+
+trait Validator[PARSED] extends Logging {
+
+  def validate: Validated[Seq[MtdError], PARSED]
+
+  def validateAndWrapResult()(implicit correlationId: String): Either[ErrorWrapper, PARSED] = {
+    validate match {
+      case Valid(parsed) =>
+        logger.info(s"Validation successful for the request with CorrelationId: $correlationId")
+        Right(parsed)
+
+      case Invalid(errs) =>
+        combineErrors(errs) match {
+          case err :: Nil =>
+            logger.warn(s"Validation failed with ${err.code} error for the request with CorrelationId: $correlationId")
+            Left(ErrorWrapper(correlationId, err, None))
+
+          case errs =>
+            logger.warn(s"Validation failed with ${errs.map(_.code).mkString(",")} error for the request with CorrelationId: $correlationId")
+            Left(ErrorWrapper(correlationId, BadRequestError, Some(errs)))
+        }
+    }
+  }
+
+  private def combineErrors(errors: Seq[MtdError]): Seq[MtdError] = {
+    errors
+      .groupBy(_.message)
+      .map { case (_, errors) =>
+        val baseError = errors.head.copy(paths = Some(Seq.empty[String]))
+
+        errors.fold(baseError)((error1, error2) => {
+          val paths: Option[Seq[String]] = for {
+            error1Paths <- error1.paths
+            error2Paths <- error2.paths
+          } yield {
+            error1Paths ++ error2Paths
+          }
+          error1.copy(paths = paths)
+        })
+      }
+      .toList
+      .sortBy(_.code)
```

Comment:
Do we need to sort by code? Pretty much every error is a 400. Except for 404 errors of course, but I don't see vendors ever getting a 400 and a 404 error at the same time.
```suggestion

```

Reply:
We were getting some 400 errors in different order (e.g. FORMAT_TAX_YEAR before FORMAT_NINO) than expected causing certain tests to fail. This was added to rectify that.

Reply:
Interesting. I'm not aware of any reason why our tests should demand errors are returned in a specific order. Have we considered telling the tests to stop caring about the error order? Just curious, I don't think we should do that right now as it will probs take a while, but longer term it might be worth the investment...

Reply:
It can either be solved by making the ordering predictive (as in the change to the common Validator code here), or by updating all the relevant tests. As there's no downside for the vendor either way, a solution that involves a minor change in one place rather than in all or nearly all the tests, seems better.

Reply:
Agree it makes sense to skip re-writing lots of old tests. 

However from now on when we're writing new tests, it seems like we should avoid unintentionally adding an unnecessary "ordering" condition if possible?